pkg/database/batch.go

```diff
@@ -0,0 +1,118 @@
+// Copyright Contributors to the Open Cluster Management project
+
+package database
+
+import (
+	"context"
+	"sync"
+
+	pgx "github.com/jackc/pgx/v4"
+	"github.com/stolostron/search-indexer/pkg/model"
+	"k8s.io/klog/v2"
+)
+
+// This is a wrapper for pgx.Batch
+// We need this because pgx.Batch doesn't provide a way to retry batches with errors.
+// Also adds reporting of errors.
+
+type batchItem struct {
+	query  string
+	args   []interface{}
+	action string // Used to report errors.
+	uid    string // Used to report errors.
+}
+
+type batchWithRetry struct {
+	items        []batchItem
+	dao          *DAO
+	wg           *sync.WaitGroup
+	syncResponse *model.SyncResponse
+}
+
+func NewBatchWithRetry(dao *DAO, syncResponse *model.SyncResponse) batchWithRetry {
+	batch := batchWithRetry{
+		items:        make([]batchItem, 0),
+		wg:           &sync.WaitGroup{},
+		dao:          dao,
+		syncResponse: syncResponse,
+	}
+	return batch
+}
+
+func (b *batchWithRetry) Queue(item batchItem) {
+	b.items = append(b.items, item)
+
+	if len(b.items) >= b.dao.batchSize {
```

Comment:
What happens here if `len(b.items) < b.dao.batchSize`? Is this workflow only for `>= b.dao.batchSize` condition?

Reply:
The logic here decides when to send the batch. We add batchItems to the queue (b.items) until we reach the batchSize, at that point we send the batch and clear the queue. I'll add this comment in the code.

The `flush()` function below is used to clear the queue if there's any remaining items that haven't been processed.

Reply:
So nothing happens when `len(b.items) < b.dao.batchSize` because we need to wait for more items to get added to the queue.